s3/pkg/service/multipart.go

```diff
@@ -444,8 +444,202 @@ func (s *s3Service) AbortMultipartUpload(ctx context.Context, in *pb.AbortMultip
 	return nil
 }
 
+func (s *s3Service) CopyObjPart(ctx context.Context, in *pb.CopyObjPartRequest, out *pb.CopyObjPartResponse) error {
+	log.Info("CopyObjPart is called in s3 service.")
+	var err error
+	defer func() {
+		out.ErrorCode = GetErrCode(err)
+	}()
+
+	srcBucketName := in.SourceBucket
+	srcObjectName := in.SourceObject
+	targetBucketName := in.TargetBucket
+	targetObjectName := in.TargetObject
+	uploadId := in.UploadID
+	partId := in.PartID
+	size := in.ReadLength
+	offset := in.ReadOffset
+
+	srcBucket, err := s.MetaStorage.GetBucket(ctx, srcBucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	srcObject, err := s.MetaStorage.GetObject(ctx, srcBucketName, srcObjectName, true)
+	if err != nil {
+		log.Errorln("failed to get object info from meta storage. err:", err)
+		return err
+	}
+	_, _, err = CheckRights(ctx, srcObject.TenantId)
+	if err != nil {
+		log.Errorf("no rights to access the source object[%s]\n", srcObject.ObjectKey)
+		return nil
+	}
+	backendName := srcBucket.DefaultLocation
+	if srcObject.Location != "" {
+		backendName = srcObject.Location
+	}
+	srcBackend, err := utils.GetBackend(ctx, s.backendClient, backendName)
+	if err != nil {
+		log.Errorln("failed to get backend client with err:", err)
+		return err
+	}
+	srcSd, err := driver.CreateStorageDriver(srcBackend.Type, srcBackend)
+	if err != nil {
+		log.Errorln("failed to create storage driver. err:", err)
+		return err
+	}
+	targetBucket, err := s.MetaStorage.GetBucket(ctx, targetBucketName, true)
+	if err != nil {
+		log.Errorln("get bucket failed with err:", err)
+		return err
+	}
+	isAdmin, tenantId, err := util.GetCredentialFromCtx(ctx)
+	if err != nil && isAdmin == false {
+		log.Error("get tenant id failed")
+		err = ErrInternalError
+		return err
+	}
+	if !isAdmin {
```

Comment:
Do we need to check ACL of both target bucket and source bucket, and source object maybe?

Reply:
@sfzeng   I added common function to check get object acl at the pr of "valid acl rules", so the source object acl check will be added after it is merged.  I add a new issue to track.

Reply:
OK, thanks.